controllers/opentelemetrycollector_controller.go

```diff
@@ -114,6 +123,124 @@ func (r *OpenTelemetryCollectorReconciler) removeRouteTask(ora autodetect.OpenSh
+			"object_kind", obj.GetObjectKind(),
+		)
+
+		if isNamespaceScoped(obj) {
+			if setErr := ctrl.SetControllerReference(&params.Instance, obj, params.Scheme); setErr != nil {
+				l.Error(setErr, "failed to set controller owner reference to resource")
+				errs = append(errs, setErr)
+				continue
+			}
+		}
+
+		desired := obj.DeepCopyObject().(client.Object)
+		mutateFn := manifests.MutateFuncFor(obj, desired)
+		op, crudErr := ctrl.CreateOrUpdate(ctx, r.Client, obj, mutateFn)
+		if crudErr != nil && errors.Is(crudErr, manifests.ImmutableChangeErr) {
+			l.Error(crudErr, "detected immutable field change, trying to delete, new object will be created on next reconcile", "obj", obj.GetName())
+			delErr := r.Client.Delete(ctx, obj)
+			if delErr != nil {
+				return delErr
+			}
+			continue
+		} else if crudErr != nil {
+			l.Error(crudErr, "failed to configure resource")
+			errs = append(errs, crudErr)
+			continue
+		}
+
+		l.V(1).Info(fmt.Sprintf("resource has been %s", op))
+
+		delete(pruneObjects, obj.GetUID())
+	}
+	if len(errs) > 0 {
+		return fmt.Errorf("failed to create objects for Collector %s: %w", params.Instance.GetName(), errors.Join(errs...))
+	}
+
+	// Prune owned objects in the cluster which are not managed anymore.
+	var pruneErrs []error
+	for _, obj := range pruneObjects {
+		l := r.log.WithValues(
+			"object_name", obj.GetName(),
+			"object_kind", obj.GetObjectKind(),
+		)
+		l.Info("pruning unmanaged resource")
+
+		err = r.Delete(ctx, obj)
+		if err != nil {
+			l.Error(err, "failed to delete resource")
+			pruneErrs = append(pruneErrs, err)
+		}
+	}
+	if len(pruneErrs) > 0 {
+		return fmt.Errorf("failed to prune objects of Collector %s: %w", params.Instance.GetName(), errors.Join(pruneErrs...))
+	}
+	return nil
+}
+
+func isNamespaceScoped(obj client.Object) bool {
+	switch obj.(type) {
+	case *rbacv1.ClusterRole, *rbacv1.ClusterRoleBinding:
+		return false
+	default:
+		return true
+	}
+}
+
+func (r *OpenTelemetryCollectorReconciler) findObjectsOwnedByOtelOperator(ctx context.Context, instance v1alpha1.OpenTelemetryCollector) (map[types.UID]client.Object, error) {
+	ownedObjects := map[types.UID]client.Object{}
+	listOps := &client.ListOptions{
+		Namespace:     instance.GetNamespace(),
+		LabelSelector: labels.SelectorFromSet(collector.SelectorLabels(instance)),
+	}
+	ingressList := &networkingv1.IngressList{}
+	err := r.List(ctx, ingressList, listOps)
+	if err != nil {
+		return nil, fmt.Errorf("error listing ingress: %w", err)
+	}
+	for i := range ingressList.Items {
+		ownedObjects[ingressList.Items[i].GetUID()] = &ingressList.Items[i]
+	}
+	if instance.Spec.Ingress.Type == v1alpha1.IngressTypeRoute {
+		routesList := &routev1.RouteList{}
+		err = r.List(ctx, routesList, listOps)
+		if err != nil {
+			return nil, fmt.Errorf("error listing routes: %w", err)
+		}
+		for i := range routesList.Items {
+			ownedObjects[routesList.Items[i].GetUID()] = &routesList.Items[i]
+		}
+	}
+	return ownedObjects, nil
```

Comment:
Am I misunderstanding something, or does this function only return Ingresses and Routes? Shouldn't it return all owned objects? Is there a test for this?

Reply:
so we do this only for objects that can be conditionally created by the operator based on a flag. This was pulled from [the tempo-operator](https://github.com/grafana/tempo-operator/blob/main/controllers/tempo/tempostack_create_or_update.go#L223) but on second look, im also not sure we need to do this. Let me run some tests.

Reply:
And with this change, that happens in the `ctrl.SetControllerReference` call in the reconciliation loop, right? In that case, we shouldn't need to do anything special for routes and ingressses.

Reply:
yeah i agree, im going to test removing this logic now :) if only my kind cluster liked me